docs/specs/app-initiated-sso-to-web.md

```diff
@@ -0,0 +1,179 @@
+
+### Token Exchange Request
+
+When a mobile app wants to initiate SSO to web, it makes a /token endpoint request to obtain a device browser session token with the following parameters:
+
+- `grant_type`: Must be `urn:ietf:params:oauth:grant-type:token-exchange`
+- `subject_token`: The id_token obtained by the mobile app
+- `subject_token_type`: Must be `urn:ietf:params:oauth:token-type:id_token`
+- `actor_token`: The device_secret obtained by the mobile app
+- `actor_token_type`: Must be `urn:x-oath:params:oauth:token-type:device-secret`
+- `requested_token_type`: Optional. If given, the value must be `urn:authgear:params:oauth:token-type:device-browser-session-token`.
+
+Here is an example:
+
+```
+POST /token HTTP/1.1
+Host: auth.example.com
+...
+grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange
+&subject_token=<id_token>
+&subject_token_type=urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aid-token
+&actor_token=95twdf3w4y6wvftw35634t
+&actor_token_type=urn%3Ax-oath%3Aparams%3Aoauth%3Atoken-type%3Adevice-secret
+&requested_token_type=urn%3Aauthgear%3Aparams%3Aoauth%3Atoken-type%3Adevice-browser-session-token
+```
+
+### Token Exchange Response
+
+The Token Exchange response for this profile has the following characteristics:
+
+- `device_secret`: Required. A new device secret must be returned. The mobile app should update its device secret with the new secret.
+- `x_device_browser_session_token`: Required. The device browser session token which can then be used to exchange for a browser cookie.
+- `expires_in`: Required. When the device browser session token expires.
+
+Here is an example:
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Cache-Control: no-store
+
+{
+  "device_secret":"casdfgarfgasdfg",
+  "x_device_browser_session_token": "zxczxczxc",
+  "expires_in":300
+}
+```
+
+## Exchanging for browser cookie of a session with the authorization endpoint
+
+A web browser can open the /authorize endpoint to exchange for a valid browser session cookie with the following parameters:
+
+- `client_id`: Required. The client_id of the obtained session.
+- `id_token_hint`: Required. The id_token obtained by the mobile app.
+- `x_device_browser_session_token`: Required. The device browser session token obtained from token exchange response.
+- `scope`: Optional. Scope of the session of the browser session cookie. If not provided, it will be same as the scope of the session which generates this browser session cookie (i.e. The offline grant of the mobile app).
+- `prompt`: Optional. If provided, must be `none`.
+- `response_type`: Required. Must be one of the following values:
+  - `session`: A long lived token will be set in the cookie, which shares the same lifetime of the refresh_token. However, it will not live longer than an IDP session.
+  - `token`: A short lived token will be set in the cookie, which shares the same lifetime of an access_token.
```

Comment:
Implementation-wise is to better to require the client to specify it is performing App-initiated SSO to Web. So `response_type` should either be `app_initiated_sso_to_web session` or `app_initiated_sso_to_web token`.